<a href="#" id="hw"></a>
<div style="background: #000;
            color: #FFF;
            margin: 0px;
                padding: 10px 0px 20px 0px;
            text-align: center; 
                ">
    <h1 style="padding: 0px 0px 0px 0px;">Week 15 Lab</h1>
</div>

# Exercise 1

a) Create a dictionary with fake data and use it to create a dataframe

In [18]:
import pandas as pd
import numpy as np

In [5]:
df = pd.DataFrame(
        {
            "State": [
                "New York",
                "California",
                "Wisconsin",
            ],
            "Capital": [
                "Albany",
                "Sacramento",
                "Madison"
            ],
       })

df

State     Capital
0    New York      Albany
1  California  Sacramento
2   Wisconsin     Madison

b)Create a list of values and use it to create a series.

In [7]:
primes = pd.Series([2, 3, 5, 101, 23], name="Prime")

In [8]:
primes

0      2
1      3
2      5
3    101
4     23
Name: Prime, dtype: int64

c) Using either the dataframe or the series above, output the count, mean and standard deviation of it.

In [9]:
primes.describe()

count      5.000000
mean      26.800000
std       42.357998
min        2.000000
25%        3.000000
50%        5.000000
75%       23.000000
max      101.000000
Name: Prime, dtype: float64

# Exercise 2

For the next parts of today's lab we will be using the `grades.csv` file in `/datasets`. 

In [11]:
# initial imports

import pandas as pd

df = pd.read_csv("../datasets/grades.csv")

a) Output the first 5 rows of the dataframe

In [12]:
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

b) Output the last 5 rows of the dataframe

In [13]:
df.tail()

gender race/ethnicity parental level of education         lunch  \
995  female        group E             master's degree      standard   
996    male        group C                 high school  free/reduced   
997  female        group C                 high school  free/reduced   
998  female        group D                some college      standard   
999  female        group D                some college  free/reduced   

    test preparation course  math score  reading score  writing score  
995               completed          88             99             95  
996                    none          62             55             55  
997               completed          59             71             65  
998               completed          68             78             77  
999                    none          77             86             86

c) What is the shape of this dataframe?

In [15]:
df.shape

(1000, 8)

d) How many columns are there with no values? How much memory does this df use?

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


There are 0 coluns with no values. The dataframe uses 62.6 KB of memory.

e) Create a new column called "total score" that is the sum of the math, reading and writing score. Output the dataframe

In [17]:
df['total score'] = df['math score'] + df['reading score'] + df['writing score']
df

gender race/ethnicity parental level of education         lunch  \
0    female        group B           bachelor's degree      standard   
1    female        group C                some college      standard   
2    female        group B             master's degree      standard   
3      male        group A          associate's degree  free/reduced   
4      male        group C                some college      standard   
..      ...            ...                         ...           ...   
995  female        group E             master's degree      standard   
996    male        group C                 high school  free/reduced   
997  female        group C                 high school  free/reduced   
998  female        group D                some college      standard   
999  female        group D                some college  free/reduced   

    test preparation course  math score  reading score  writing score  \
0                      none          72             72             74   
1                 completed          69             90             88   
2                      none          90             95             93   
3                      none          47             57             44   
4                      none          76             78             75   
..                      ...         ...            ...            ...   
995               completed          88             99             95   
996                    none          62             55             55   
997               completed          59             71             65   
998               completed          68             78             77   
999                    none          77             86             86   

     total score  
0            218  
1            247  
2            278  
3            148  
4            229  
..           ...  
995          282  
996          172  
997          195  
998          223  
999          249  

[1000 rows x 9 columns]

f) Create a new column called "average score" that is the average of the math, reading, and writing scores. Output the dataframe.

In [22]:
df['average score'] = df[['math score', 'reading score', 'writing score']].mean(axis=1)
df

gender race/ethnicity parental level of education         lunch  \
0    female        group B           bachelor's degree      standard   
1    female        group C                some college      standard   
2    female        group B             master's degree      standard   
3      male        group A          associate's degree  free/reduced   
4      male        group C                some college      standard   
..      ...            ...                         ...           ...   
995  female        group E             master's degree      standard   
996    male        group C                 high school  free/reduced   
997  female        group C                 high school  free/reduced   
998  female        group D                some college      standard   
999  female        group D                some college  free/reduced   

    test preparation course  math score  reading score  writing score  \
0                      none          72             72             74   
1                 completed          69             90             88   
2                      none          90             95             93   
3                      none          47             57             44   
4                      none          76             78             75   
..                      ...         ...            ...            ...   
995               completed          88             99             95   
996                    none          62             55             55   
997               completed          59             71             65   
998               completed          68             78             77   
999                    none          77             86             86   

     total score  average score  
0            218      72.666667  
1            247      82.333333  
2            278      92.666667  
3            148      49.333333  
4            229      76.333333  
..           ...            ...  
995          282      94.000000  
996          172      57.333333  
997          195      65.000000  
998          223      74.333333  
999          249      83.000000  

[1000 rows x 10 columns]

g) What is the average `total score` for each gender? What is the average `avg score` for each gender?

**Average total score for each gender**

In [23]:
df.groupby('gender')['total score'].mean()

gender
female    208.708494
male      197.512448
Name: total score, dtype: float64

**Average avg score for each gender**

In [24]:
df.groupby('gender')['average score'].mean()

gender
female    69.569498
male      65.837483
Name: average score, dtype: float64

h) What is the average math score for each category of lunch? What about reading, writing, total and avg score

**Average scores (math, reading, writing, total, avg) for each category of lunch**

In [27]:
df.groupby('lunch')[['math score', 'reading score', 'writing score', 'total score', 'average score']].mean()

math score  reading score  writing score  total score  \
lunch                                                                 
free/reduced   58.921127      64.653521      63.022535   186.597183   
standard       70.034109      71.654264      70.823256   212.511628   

              average score  
lunch                        
free/reduced      62.199061  
standard          70.837209

i) What is the descriptive statistics for each category of `parental level of education`? What about reading, writing, total and avg score

In [32]:
df.groupby('parental level of education')[['math score', 'reading score', 'writing score', 'total score', 'average score']].describe()

math score                                     \
                                 count       mean        std   min    25%   
parental level of education                                                 
associate's degree               222.0  67.882883  15.112093  26.0  57.00   
bachelor's degree                118.0  69.389831  14.943789  29.0  61.00   
high school                      196.0  62.137755  14.539651   8.0  53.75   
master's degree                   59.0  69.745763  15.153915  40.0  55.50   
some college                     226.0  67.128319  14.312897  19.0  59.00   
some high school                 179.0  63.497207  15.927989   0.0  53.00   

                                               reading score             ...  \
                              50%   75%    max         count       mean  ...   
parental level of education                                              ...   
associate's degree           67.0  80.0  100.0         222.0  70.927928  ...   
bachelor's degree            68.0  79.0  100.0         118.0  73.000000  ...   
high school                  63.0  72.0   99.0         196.0  64.704082  ...   
master's degree              73.0  81.0   95.0          59.0  75.372881  ...   
some college                 67.5  76.0  100.0         226.0  69.460177  ...   
some high school             65.0  74.0   97.0         179.0  66.938547  ...   

                            total score        average score             \
                                    75%    max         count       mean   
parental level of education                                               
associate's degree                237.0  300.0         222.0  69.569069   
bachelor's degree                 242.0  300.0         118.0  71.923729   
high school                       218.0  287.0         196.0  63.096939   
master's degree                   256.5  293.0          59.0  73.598870   
some college                      234.0  297.0         226.0  68.476401   
some high school                  229.5  297.0         179.0  65.108007   

                                                                         \
                                   std        min        25%        50%   
parental level of education                                               
associate's degree           13.670914  31.666667  58.666667  69.666667   
bachelor's degree            13.946609  39.000000  64.083333  71.166667   
high school                  13.510583  18.333333  53.916667  65.000000   
master's degree              13.601017  44.666667  63.166667  73.333333   
some college                 13.710974  23.333333  60.000000  68.666667   
some high school             14.984078   9.000000  55.666667  66.666667   

                                                    
                                   75%         max  
parental level of education                         
associate's degree           79.000000  100.000000  
bachelor's degree            80.666667  100.000000  
high school                  72.666667   95.666667  
master's degree              85.500000   97.666667  
some college                 78.000000   99.000000  
some high school             76.500000   99.000000  

[6 rows x 40 columns]

j) What are the average scores (for all score columns) for each category of `test preparation course`? 

In [34]:
df.groupby('test preparation course')[['math score', 'reading score', 'writing score', 'total score', 'average score']].mean()

math score  reading score  writing score  \
test preparation course                                             
completed                 69.695531      73.893855      74.418994   
none                      64.077882      66.534268      64.504673   

                         total score  average score  
test preparation course                              
completed                 218.008380      72.669460  
none                      195.116822      65.038941

k) Output only the score columns

In [35]:
df[['math score', 'reading score', 'writing score', 'total score', 'average score']]

math score  reading score  writing score  total score  average score
0            72             72             74          218      72.666667
1            69             90             88          247      82.333333
2            90             95             93          278      92.666667
3            47             57             44          148      49.333333
4            76             78             75          229      76.333333
..          ...            ...            ...          ...            ...
995          88             99             95          282      94.000000
996          62             55             55          172      57.333333
997          59             71             65          195      65.000000
998          68             78             77          223      74.333333
999          77             86             86          249      83.000000

[1000 rows x 5 columns]

l) What rows have `avg scores` lower than 80? What percent of the total is that? (try to calculate the last part programmatically)

In [41]:
less_than_80 = df[df['average score'] < 80]
less_than_80

gender race/ethnicity parental level of education         lunch  \
0    female        group B           bachelor's degree      standard   
3      male        group A          associate's degree  free/reduced   
4      male        group C                some college      standard   
5    female        group B          associate's degree      standard   
7      male        group B                some college  free/reduced   
..      ...            ...                         ...           ...   
993  female        group D           bachelor's degree  free/reduced   
994    male        group A                 high school      standard   
996    male        group C                 high school  free/reduced   
997  female        group C                 high school  free/reduced   
998  female        group D                some college      standard   

    test preparation course  math score  reading score  writing score  \
0                      none          72             72             74   
3                      none          47             57             44   
4                      none          76             78             75   
5                      none          71             83             78   
7                      none          40             43             39   
..                      ...         ...            ...            ...   
993                    none          62             72             74   
994                    none          63             63             62   
996                    none          62             55             55   
997               completed          59             71             65   
998               completed          68             78             77   

     total score  average score  
0            218      72.666667  
3            148      49.333333  
4            229      76.333333  
5            232      77.333333  
7            122      40.666667  
..           ...            ...  
993          208      69.333333  
994          188      62.666667  
996          172      57.333333  
997          195      65.000000  
998          223      74.333333  

[802 rows x 10 columns]

**Percent of scores that have average less than 80**

In [45]:
print(less_than_80.shape[0] / df.shape[0] * 100, '%')

80.2 %


m) What are the `avg score`, `gender` and `lunch` columns for all students with avg scores above 90. Sort it in descending order by `avg score`. Store it to a new dataframe and then output the new dataframe.

In [47]:
new_df = df.loc[df['average score'] > 90, ['average score', 'gender', 'lunch']].sort_values("average score", ascending=False)
new_df

average score  gender         lunch
458     100.000000  female      standard
962     100.000000  female      standard
916     100.000000    male      standard
114      99.666667  female      standard
179      99.000000  female      standard
712      99.000000  female      standard
625      98.666667    male      standard
165      98.666667  female      standard
903      97.666667  female  free/reduced
685      97.666667  female      standard
149      97.666667    male  free/reduced
957      97.333333  female      standard
566      97.333333  female  free/reduced
594      97.000000  female      standard
717      97.000000  female      standard
855      96.666667  female      standard
970      96.333333  female      standard
546      96.333333  female      standard
451      96.333333  female      standard
886      96.000000  female      standard
106      95.666667  female      standard
403      95.666667  female      standard
995      94.000000  female      standard
381      94.000000    male      standard
623      94.000000    male      standard
263      94.000000  female      standard
864      93.666667    male      standard
979      93.333333  female      standard
377      93.333333  female  free/reduced
571      93.000000    male      standard
919      92.666667    male      standard
2        92.666667  female      standard
562      92.666667    male      standard
229      92.333333  female      standard
503      92.000000  female      standard
802      92.000000  female      standard
934      91.666667    male      standard
268      91.666667  female      standard
612      91.666667    male      standard
539      91.666667    male      standard
6        91.666667  female      standard
373      91.666667  female      standard
580      91.333333  female      standard
122      91.333333  female      standard
104      91.333333    male      standard
501      91.000000  female      standard
121      90.666667    male      standard
474      90.333333  female      standard
334      90.333333  female      standard
755      90.333333  female      standard

m) How many rows are in the new dataframe? What percent of the new dataframe are female? What percent receive free/reduced lunch?

**Rows in new dataframe**

In [48]:
new_df.shape[0]

50

**Percent of new_df that is female**

In [63]:
(new_df['gender'].value_counts() / new_df['gender'].count()) * 100

female    72.0
male      28.0
Name: gender, dtype: float64

**Percent of new_df receiving free/reduced lunch**

In [64]:
(new_df['lunch'].value_counts() / new_df['lunch'].count()) * 100

standard        92.0
free/reduced     8.0
Name: lunch, dtype: float64

**ANSWER: 72% of the new dataframe is female while 8% receives free/reduced lunch**

n) Using the original dataframe, what is the `avg score` for each of the `gender`, `lunch` and `parental education level` column combinations?

In [72]:
df.groupby(['gender', 'lunch', 'parental level of education'])['average score'].mean()

gender  lunch         parental level of education
female  free/reduced  associate's degree             65.907407
                      bachelor's degree              71.060606
                      high school                    58.405405
                      master's degree                67.333333
                      some college                   65.378788
                      some high school               56.425926
        standard      associate's degree             73.470833
                      bachelor's degree              76.609756
                      high school                    68.865497
                      master's degree                77.666667
                      some college                   74.346847
                      some high school               71.545455
male    free/reduced  associate's degree             65.008130
                      bachelor's degree              63.045455
                      high school                    56.222222
     

o) Sort the original dataframe by `gender` and `lunch` in descending order

In [75]:
df.sort_values(by=['gender', 'lunch'], ascending=False)

gender race/ethnicity parental level of education         lunch  \
4      male        group C                some college      standard   
10     male        group C          associate's degree      standard   
11     male        group D          associate's degree      standard   
13     male        group A                some college      standard   
16     male        group C                 high school      standard   
..      ...            ...                         ...           ...   
989  female        group D                some college  free/reduced   
992  female        group D          associate's degree  free/reduced   
993  female        group D           bachelor's degree  free/reduced   
997  female        group C                 high school  free/reduced   
999  female        group D                some college  free/reduced   

    test preparation course  math score  reading score  writing score  \
4                      none          76             78             75   
10                     none          58             54             52   
11                     none          40             52             43   
13                completed          78             72             70   
16                     none          88             89             86   
..                      ...         ...            ...            ...   
989               completed          67             86             83   
992                    none          55             76             76   
993                    none          62             72             74   
997               completed          59             71             65   
999                    none          77             86             86   

     total score  average score  
4            229      76.333333  
10           164      54.666667  
11           135      45.000000  
13           220      73.333333  
16           263      87.666667  
..           ...            ...  
989          236      78.666667  
992          207      69.000000  
993          208      69.333333  
997          195      65.000000  
999          249      83.000000  

[1000 rows x 10 columns]

p) Based on your exercises above, what sort of inferences can you make so far? What would you want to look further into? What factor seems to affect `avg` score the most? Explain your answer.